File: Reqnroll/Chapter13/13.3-IntroduceAndReferPattern/After/WIMP.Specs/Support/CustomParameterTypes.cs
using Reqnroll;

namespace WIMP.Specs.Support;

[Binding]
public class CustomParameterTypes(OrderingContext orderingContext)
{
    [StepArgumentTransformation("the order|the placed order|the new order", Name = "order")]
    public int ConvertOrder()
    {
        return orderingContext.CurrentOrderNo ??
            throw new InvalidOperationException("No current order");
    }
}

File: CucumberJVM/Chapter13/13.3-IntroduceAndReferPattern/After/src/test/java/com/wimp/app/specs/support/CustomParameterTypes.java
package com.wimp.app.specs.support;

import com.wimp.app.models.*;
import io.cucumber.java.ParameterType;

import java.util.Optional;

public class CustomParameterTypes {

    private final OrderingContext orderingContext;

    public CustomParameterTypes(OrderingContext orderingContext) {
        this.orderingContext = orderingContext;
    }

    @ParameterType("the order|the placed order|the new order")
    public int order(String value) {
        return Optional.ofNullable(orderingContext.getCurrentOrderNo())
            .orElseThrow(() -> new RuntimeException("No current order"));
    }
}
